templates/commands/render.go

```diff
@@ -235,24 +275,237 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
 		return err
 	}
 
-	_ = templateDir // TODO: add template rendering logic
+	spec, err := loadSpecFile(rp.fs, templateDir, r.flagSpec)
+	if err != nil {
+		return err
+	}
+
+	scratchDir, err := rp.tempDirNamer("scratch")
+	if err != nil {
+		return err
+	}
+	if err := rp.fs.MkdirAll(scratchDir, 0o700); err != nil {
+		return fmt.Errorf("MkdirAll(): %w", err)
+	}
 
+	if err := executeSpec(ctx, spec, &stepParams{
+		inputs:      r.flagInputs,
+		fs:          rp.fs,
+		scratchDir:  scratchDir,
+		stdout:      rp.stdout,
+		templateDir: templateDir,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func executeSpec(ctx context.Context, spec *model.Spec, sp *stepParams) error {
+	for _, step := range spec.Steps {
+		if err := executeOneStep(ctx, step, sp); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
+type stepParams struct {
+	fs          renderFS
+	inputs      map[string]string
+	scratchDir  string
+	stdout      io.Writer
+	templateDir string
+}
+
+func executeOneStep(ctx context.Context, step *model.Step, sp *stepParams) error {
+	switch {
+	case step.Print != nil:
+		return actionPrint(ctx, step.Print, sp)
+	case step.Include != nil:
+		return actionInclude(ctx, step.Include, sp)
+	case step.RegexReplace != nil:
+		return actionRegexReplace(ctx, step.RegexReplace, sp)
+	case step.StringReplace != nil:
+		return actionStringReplace(ctx, step.StringReplace, sp)
+	case step.GoTemplate != nil:
+		return actionGoTemplate(ctx, step.GoTemplate, sp)
+	default:
+		return fmt.Errorf("internal error: unknown step action type %q", step.Action.Val)
```

Comment:
Is this an _internal_ error? Or is this is a user error?

Reply:
It's an internal error. If the user screwed up, we would have caught it in the parsing step. This error means that the parser knows about an action type but the renderer doesn't. So it will "never" happen.